src/System.Private.ServiceModel/src/System/ServiceModel/Security/SecurityProtocolFactory.cs

```diff
@@ -686,6 +693,31 @@ public virtual void OnClose(TimeSpan timeout)
             }
         }
 
+        public virtual Task OnCloseAsync(TimeSpan timeout)
+        {
+            return OnCloseAsyncInternal(timeout);
+        }
+
+        private async Task OnCloseAsyncInternal(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            if (!_actAsInitiator)
+            {
+                foreach (SupportingTokenAuthenticatorSpecification spec in _channelSupportingTokenAuthenticatorSpecification)
+                {
+                    await SecurityUtils.CloseTokenAuthenticatorIfRequiredAsync(spec.TokenAuthenticator, timeoutHelper.RemainingTime());
+                }
+                foreach (string action in _scopedSupportingTokenAuthenticatorSpecification.Keys)
```

Comment:
I don't see the keys being used - might it make more sense to just enumerate through _scopedSupportingTokenAuthenticatorSpecification.Values to get the `ICollection<SupportingTokenAuthenticatorSpecification>` rather than indexing based on action?